api/users/views.py

```diff
@@ -655,3 +658,96 @@ def post(self, request, *args, **kwargs):
+
+    required_read_scopes = [CoreScopes.USER_SETTINGS_READ]
+    required_write_scopes = [CoreScopes.USER_SETTINGS_WRITE]
+
+    view_category = 'users'
+    view_name = 'user-emails'
+
+    serializer_class = UserEmailsSerializer
+
+    # overrides ListAPIViewa
+    def get_queryset(self):
+        user = self.get_user()
+        serialized_emails = []
+        for email in user.emails.all():
+            primary = email.address == user.username
+            hashed_id = hashids.encode(email.id)
+            serialized_email = UserEmail(email_id=hashed_id, address=email.address, confirmed=True, primary=primary)
+            serialized_emails.append(serialized_email)
+        email_verifications = user.email_verifications or []
+        for token in email_verifications:
+            detail = user.email_verifications[token]
+            serialized_unconfirmed_email = UserEmail(email_id=token, address=detail['email'], confirmed=detail['confirmed'], primary=False)
+            serialized_emails.append(serialized_unconfirmed_email)
+
+        return serialized_emails
+
+
+class UserEmailsDetail(JSONAPIBaseView, generics.RetrieveUpdateDestroyAPIView, UserMixin):
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        base_permissions.TokenHasScope,
+        CurrentUser,
+    )
+
+    required_read_scopes = [CoreScopes.USER_SETTINGS_READ]
+    required_write_scopes = [CoreScopes.USER_SETTINGS_WRITE]
+
+    view_category = 'users'
+    view_name = 'user-email-detail'
+
+    serializer_class = UserEmailsSerializer
+
+    # Overrides RetrieveUpdateDestroyAPIView
+    def get_object(self):
+        email_id = self.kwargs['email_id']
+        user = self.get_user()
+
+        # check to see if it's a confirmed email with hashed id
+        decoded_id = hashids.decode(email_id)
+        if decoded_id:
+            try:
+                email = user.emails.get(id=decoded_id[0])
+                primary = email.address == user.username
+                address = email.address
+                confirmed = True
+            except Email.DoesNotExist:
+                email = None
+
+        # check to see if it's an unconfirmed email with a token
+        elif user.unconfirmed_emails:
+            try:
+                email = user.email_verifications[email_id]
+                address = email['email']
+                confirmed = email['confirmed']
+                primary = False
+            except KeyError:
+                email = None
+
+        if not email:
+            raise NotFound
+
+        return UserEmail(email_id=email_id, address=address, confirmed=confirmed, primary=primary)
+
+    # Overrides RetrieveUpdateDestroyAPIView
+    def perform_destroy(self, instance):
+        user = self.get_user()
+        email = instance.address
+        if instance.confirmed:
+            try:
+                user.remove_email(email)
+            except PermissionsError as e:
+                raise ValidationError(e.message)
```

Comment:
Minor: To ease the transition to Python 3, you can change `e.message` to `e.args[0]`. Built-in `Exception` no longer has the `message` attribute.